spinalcordtoolbox/csa_pmj.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python
+# -*- coding: utf-8
+# Functions to get distance from PMJ for processing segmentation data
+# Author: Sandrine Bédard
+import logging
+import numpy as np
+
+from spinalcordtoolbox.image import Image
+from spinalcordtoolbox.centerline.core import get_centerline
+from spinalcordtoolbox.centerline import curve_fitting
+from spinalcordtoolbox.resampling import resample_nib
+
+logger = logging.getLogger(__name__)
+
+NEAR_ZERO_THRESHOLD = 1e-6
+
+
+def get_slices_for_pmj_distance(segmentation, pmj, distance, extent, param_centerline=None, verbose=1):
+    """
+    Compute distance from PMJ projection on centerline for all the centerline.
+    Generate mask from segmentation of the slices used to process segmentation data corresponding to a distance from PMJ projection.
+    :param segmentation: input segmentation. Could be either an Image or a file name.
+    :param pmj: label of PMJ.
+    :param distance: float: Distance from Ponto-Medullary Junction (PMJ) in mm.
+    :param param_centerline: see centerline.core.ParamCenterline()
+    :param verbose:
+    :return im_ctl:
+    :return mask:
+    :return slices:
+
+    """
+    native_orientation = Image(segmentation).orientation
+    im_seg = Image(segmentation).change_orientation('RPI')
+    im_pmj = Image(pmj).change_orientation('RPI')
+    nx, ny, nz, nt, px, py, pz, pt = im_seg.dim
+    pr = min([px, py])
+    im_segr = resample_nib(im_seg, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    im_pmjr = resample_nib(im_pmj, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    data_pmj = im_pmjr.data
+    # Update dimensions from resampled image
+    nx, ny, nz, nt, px, py, pz, pt = im_segr.dim
+
+    # Extract min and max index in Z direction
+    data_seg = im_segr.data
+    X, Y, Z = (data_seg > NEAR_ZERO_THRESHOLD).nonzero()
+    min_z_index, max_z_index = min(Z), max(Z)
+
+    # Remove top slices  | TODO: check if center of mass of top slices is close to other slices, if not, remove
+    im_segr.data[:, :, max_z_index - 4:max_z_index + 1] = 0
```

Comment:
This code deals with cases where the top slices of the segmentations are only partial, which could yield wrong estimate of the extrapolation. Example below:

with this code:
![fig_centerline_210628-125903372620_bspline](https://user-images.githubusercontent.com/2482071/123691178-494a6780-d823-11eb-94d4-ecb115056bda.png)


without this code (notice the ^106 for the Y component):
![fig_centerline_210628-150754841347_bspline](https://user-images.githubusercontent.com/2482071/123691214-55cec000-d823-11eb-969d-c99924b8b951.png)

However, I am not a big fan of this code because we could imagine cases where more than the top four slices show a not well defined segmentation. In general, extrapolation seems poorly robust so I would be in favor of avoiding it.

One possibility I'm thinking about is including the PMJ in the segmentation and then interpolating between the segmentation and the PMJ. That way we would avoid robustness issues. I'll investigate a bit and document in this thread...

Reply:
Great thanks for investigating!

> One possibility I'm thinking about is including the PMJ in the segmentation and then interpolating between the segmentation and the PMJ. 

The one thing maybe to look out for with this approach is to validate that the curvature of the spinal cord will be taken into account and not biased with the PMJ coordinate (https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3063#issuecomment-862648093)

Reply:
> The one thing maybe to look out for with this approach is to validate that the curvature of the spinal cord will be taken into account and not biased with the PMJ coordinate ([#3063 (comment)](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3063#issuecomment-862648093))

Yes, this is a very valid point. I am wondering what is the most valid approach: using the projection on an extrapolated centerline (where the extrapolation can be problematic) or the length to the PMJ (even though it is slightly displaced anteriorly compared to the centerline). My intuition says that the latter is better. This is something we could/should validate by acquiring high resolution T2w scans where we can see the nerve rootlets.

Reply:
> This is something we could/should validate by acquiring high resolution T2w scans where we can see the nerve rootlets.
yes that will definitely be interesting!